### PR TITLE
BHBC-2168: Queue Scheduler Process

### PR DESCRIPTION
**Pull request by NickPhura**
# Overview

## Links to Jira tickets

https://quartech.atlassian.net/browse/BHBC-2168

## Description of relevant changes

New queue process that is designed to be deployed on its own (own dockerfile, etc).
Will poll the database for queue records, kicking off jobs, etc.

## Testing

To insert a queue record for testing purposes, you can run the following as many times as you want.

```
  with sub as (
      insert into submission (
          source_transform_id, 
          uuid
      )
      values (
          1, 
          public.gen_random_uuid()
      )
      returning submission_id
  )
  insert into submission_job_queue (
      submission_job_queue_id,
      submission_id,
      key
  )
  values (
      (select nextval('submission_job_queue_seq')),
      (select submission_id from sub),
      'a/fake/key'
  );
```

## PR Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

### CODE REVIEW

File: api/src/queue/queue-scheduler.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+import { SubmissionJobQueueService } from '../services/submission-job-queue-service';
+import { SystemConstantService } from '../services/system-constant';
+import { getLogger } from '../utils/logger';
+import { Queue } from './queue';
+
+const defaultLog = getLogger('queue/queue-scheduler');
+
+export const QUEUE_DEFAULT_ENABLED = false;
+export const QUEUE_DEFAULT_CONCURRENCY = 4;
+export const QUEUE_DEFAULT_PERIOD = 5000; // 5 seconds
```
Comment (curtisupshall) [inline api/src/queue/queue-scheduler.ts:12]: Minor nitpick: Thoughts on adding units to these two consts? E.g. `QUEUE_DEFAULT_PERIOD_MS`

File: api/src/services/system-constant.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { IDBConnection } from '../database/db';
+import { ISystemConstant, SystemConstantRepository } from '../repositories/system-constant';
+import { DBService } from './db-service';
+
+export class SystemConstantService extends DBService {
```
Comment (curtisupshall) [inline api/src/services/system-constant.ts]: +1 for system constant service 👍

File: api/src/repositories/system-constant.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+   * @return {*}  {Promise<ISystemConstant>}
+   * @memberof SystemConstantRepository
+   */
+  async getSystemConstant(constantName: string): Promise<ISystemConstant> {
+    const knex = getKnex();
+    const queryBuilder = knex.queryBuilder().select().from('system_constant').where('constant_name', constantName);
+
+    const response = await this.connection.knex<ISystemConstant>(queryBuilder);
+
+    return response.rows[0];
```
Comment (curtisupshall) [inline api/src/repositories/system-constant.ts]: Needs case when `rows === undefined`?

Comment (curtisupshall) [reply api/src/repositories/system-constant.ts]: Same for `getSystemConstants`

File: database/src/migrations/release.0.7.0/populate_system_constant.sql [hunk 4]
```diff
@@ -6,4 +6,10 @@ insert into system_constant (constant_name, character_value, description) values
...
+-- job queue
+insert into system_constant (constant_name, character_value, description) values ('JOB_QUEUE_ENABLED', 'true', 'Controls whether or not the job queue scheduler is enabled. Set to "true" to enable the job queue schdulerF, "false" to disable it.');
+insert into system_constant (constant_name, numeric_value,   description) values ('JOB_QUEUE_CONCURRENCY', 4, 'The number of job queue processes that can run concurrently (integer > 0).');
+insert into system_constant (constant_name, numeric_value,   description) values ('JOB_QUEUE_PERIOD', 5000, 'The frequency with which the job queue scheduler will check for new job queue records (milliseconds).');
+insert into system_constant (constant_name, numeric_value,   description) values ('JOB_QUEUE_ATTEMPTS', 2, 'The total number of times a job will be attempted until it finishes successfully (integer >= 1).');
+insert into system_constant (constant_name, numeric_value,   description) values ('JOB_QUEUE_TIMEOUT', 60000, 'The maximum duration a running job can take before it is considered timed out.');
```
Comment (curtisupshall) [inline database/src/migrations/release.0.7.0/populate_system_constant.sql]: Missing units in `description` column

File: api/src/repositories/submission-job-queue-repository.ts [hunk 5]
```diff
@@ -112,4 +114,132 @@ export class SubmissionJobQueueRepository extends BaseRepository {
...
+
+    queryBuilder.orderBy('submission_job_queue_id', 'ASC');
+
+    if (concurrency) {
+      queryBuilder.limit(concurrency);
+    }
+
+    const response = await this.connection.knex<ISubmissionJobQueueRecord>(queryBuilder);
+
+    return response.rows;
```
Comment (curtisupshall) [inline api/src/repositories/submission-job-queue-repository.ts:190]: `rows === undefined`

File: api/src/repositories/system-constant.ts (hunk 3 above)
Comment (NickPhura) [reply api/src/repositories/system-constant.ts]: `rows` will always be an array, so accessing the 0th index is safe, it will just return `undefined` if there are no elements.

That being said, this function is actually never used, so I am just going to remove it.

File: api/src/queue/queue-scheduler.ts (hunk 1 above)
Comment (NickPhura) [reply api/src/queue/queue-scheduler.ts:12]: No strong feelings either way. I guess the only consideration is that we should probably do the same thing with all constants then, which might get a bit messy.

File: database/src/migrations/release.0.7.0/populate_system_constant.sql (hunk 4 above)
Comment (NickPhura) [reply database/src/migrations/release.0.7.0/populate_system_constant.sql]: Added, will wait to push as testing is happening in OpenShift atm

File: api/src/repositories/submission-job-queue-repository.ts (hunk 5 above)
Comment (NickPhura) [reply api/src/repositories/submission-job-queue-repository.ts:190]: Same as above: rows will always be an array, so we shouldn't need to check it.

**Review by anissa-agahchen — DISMISSED** [review]

**Review by anissa-agahchen — DISMISSED** [review]

**Comment by AlfredRosenthal** [comment]
Pulled our changes and did some more testing, should subsequent submissions work? It looks like the transforms are failing because of a duplicate key issue on `security_transform_submission`.

**Comment by NickPhura** [comment]
> Pulled our changes and did some more testing, should subsequent submissions work? It looks like the transforms are failing because of a duplicate key issue on `security_transform_submission`.

Hmm, yeah that seems like a bug. We do want to support subsequent submissions (end-date old records, insert new records).

**Review by AlfredRosenthal — APPROVED** [review]
🥼
